File: web/apps/app-manager/app-manager.js
/*
 * CLIRA application manager
 *
 * Fetch and install CLIRA apps from a variety of sources. Also view, update 
 * or delete installed apps.
 */

var appManager = {
    view : null,
    state : {},
    mode : null,
    appList : null,
    metaDefinition : [
            { id : "name", type : "string", label : "App name", disabled : true },
            { id : "version", type : "string", label : "Version" },
            { id : "author", type : "string", label : "Author" },
            { id : "author-email", type : "string", label : "Author email" },
            { id : "app-meta-url", type : "string", label : "Metafile URL" },
            { id : "description", type : "string", label : "App description"},
            { id : "readme", type : "string", label : "Readme URL" },
            { id : "files", type : "array", label : "App file list" },
            { id : "external", type : "array", label : "External files"}
    ],
    setError : function (msg) {
        this.state = { error : msg };
        this.view.set('controller.state', this.state);
        console.error(msg);
    },
    setSource : function (src) {
        var data = {};
        data[src] = true;
        this.setView('src', data);
    },
    setSuccess : function (data) {
        this.state = { success : data };
        this.view.set('controller.state', this.state);             
    },
    setLoading : function (msg) {
        this.state = { loading : msg };
        this.view.set('controller.state', this.state);
    },
    resetState : function () {
        this.state = {};
        this.view.set('controller.state', this.state);             
    },
    setView : function (key, value) {
        this.view.set('controller.' + key, value);
    }, 
    setMode : function (val) {
        this.mode = val;
        var mode = {};
        mode[val] = true;
        this.setView('mode', mode);
    },
    getMeta : function (appName) {
        var def = $.Deferred();
        this.doAjax({
            name : appName,
            mode : 'getMeta'
        }, function (res) {
            if (res.success && res.meta) {
                var meta = res.meta;
                var list = [];
                appManager.metaDefinition.forEach(function(v, i) {
                    var data = {};
                    if (v.type == "array" && !(v.id in meta))
                        return;

                    if (v.id in meta) {
                        data.value = meta[v.id];
                        if (res.metaCreate && v.id == "files") {
                           data.value.push(appName + ".meta (Will be created)");
                        }
                    }
                    data.name = v.id;
                    data.label = v.label;
                    data[v.type] = true;
                    if ('disabled' in v)
                        data.disabled = v.disabled;
                    list.push(data);
                });
                if (res.metaCreate) 
                    appManager.setView('metaCreate', true);
                else 
                    appManager.setView('metaCreate', false);
                appManager.setView('appName', appName);
                appManager.setView('metaList', list);
            } else if (res.error) {
                appManager.setError(res.error);
            }
            def.resolve();
        });
        return def.promise();
    },
    doAjax : function (data, onSuccessCb, onErrorCb) {
        if (!onErrorCb) {
            onErrorCb = function (xhr, stat, msg) {
                error = "AJAX request failed : ";
                if (stat == "timeout")
                    error += "Request timed out";
                else 
                    error += stat;
                appManager.setError(error);
            } 
        }
        $.ajax({
            url : '/apps/app-manager/py/app-manager.py',
            data : data,
            dataType : 'json',
            success : function (res) {
                onSuccessCb(res);
            },
            error : onErrorCb,
            timeout : 10000
        });         
    },
    showAppList : function () {
        var that = this;
        var def = $.Deferred();
        this.doAjax({mode : "listApps"}, function (res) {
            if (res.success) {
                that.appList = res.appList;
                that.setView('appList', res.appList);
                def.resolve();
            }
        });
        return def.promise();
    },
    installApp : function (formData, callback) {
        this.setLoading("Installing app...")
        this.doAjax(formData, callback);
    },
    updateApp : function (appName, callback) {
        data = {
            name : appName,
            mode : 'updateApp'
        };
        this.doAjax(data, callback);
    },
    checkForUpdates : function () {
        var that = this;
        this.appList.forEach(function (app, i) {
            if (app.meta["app-meta-url"]) {
                that.doAjax({
                    mode : "checkAppUpdate",
                    url : app.meta["app-meta-url"],
                    version : app.meta.version
                }, function (res) {
                    if (res.success) { 
                        var html = "<button id='update-" + app.name + 
                            "' class='btn btn-primary update-btn btn-sm'>" + 
                            "Update</button>" + "<label>New version " + 
                            res.updateInfo['new-version'] + " available.";
                        
                         var uptoDateHtml = "<span class='glyphicon " + 
                            "glyphicon-ok-circle'></span><b>Up-to-date</b>";
                        if (res.updateInfo['new-version']) {
                            var newVer = res.updateInfo['new-version'];
                            that.view.$(".app-status#" + app.name)
                                .removeClass('app-loading')
                                .html(html);

                            that.view.$("#update-" + app.name)
                                .click(function (e) {

                                e.preventDefault();
                                $(this).html("Updating...")
                                    .prop('disabled', true);
                                $("<div class='app-loading'></div>")
                                    .insertAfter($(this));
                                
                                that.updateApp(app.name, function (res) {
                                    if (res.success) {
                                        Ember.set(app.meta, 'version', newVer);
                                        $.clira.reloadApp(app.name)
                                            .done(function () {
                                            that.view.$(".app-status#" + 
                                                app.name)
                                                .html(uptoDateHtml)
                                                .addClass('alert alert-success');
                                        });
                                    } else if (res.error) {
                                        console.err(res.error);
                                    }
                                });
                            });
                        } else {
                            that.view.$(".app-status#" + app.name)
                                .removeClass('app-loading')
                                .html(uptoDateHtml)
                                .addClass('alert alert-success');
                        }
                    } else if (res.error) {
                        var errorHtml = "<div class='alert alert-warning'>" + 
                            "<span class='glyphicon " + 
                            "glyphicon-exclamation-sign'></span>" + res.error +
                            "</div>";
                        that.view.$(".app-status#" + app.name)
                                .removeClass('app-loading')
                                .html(errorHtml);
                    }
                });
            }
        });                
    },
    getFormData : function (ctx) {
        var data = {};
        ctx.$("input:text").each(function(i) {
            if ($(this).prop('disabled') || $(this).val() == "") return;
            data[$(this).attr('name')] = $(this).val();
        });
        return data;
    }
};
Clira.appManager = {
    selectAppSrcView : Ember.View.extend({
        render : function (buffer) {
            var options = 
                "<option value='localDisk'>Local Disk</option>" + 
                "<option value='github'>Github</option>" +
                "<option value='webServer'>Web Server</option>";
            buffer.push(options);         
        },
        didInsertElement : function () {                   
            this.$("#selectAppSrc").selectric({
                onChange : function () {
                    var src = {};
                    appManager.setSource($(this).val());
                    appManager.resetState();
                }
            }); 
        }
    }),
    formInputView : Ember.View.extend({
        didInsertElement : function () {
            that = this;
            this.$("#installApp").click(function (e) {
                e.preventDefault();
                if (appManager.state.success) {
                    appManager.resetState();
                    that.$("input:text").val("");
                    that.$("input:text:first").focus();
                    $(this).html("Install");
                    return;
                }
                var data = appManager.getFormData(that);
                data.src = appManager.view.$('#selectAppSrc').val();
                appManager.installApp(data, function (res) {
                    if (res.error) {
                        appManager.setError(res.error);
                    } else if (res.success) {
                        $.clira.reloadApp(res.appName).done(function () {
                            appManager.setSuccess({
                                name : res.appName,
                                result : 'installed'
                            });
                            that.$("#installApp").html("Install another app");
                            appManager.showAppList().done(function() {
                                appManager.checkForUpdates();
                            });
                        });
                    }
                });
            });
            this.$("#metaPath").autocomplete({
                source : "/apps/app-manager/py/list-dirs.py",
                minLength : 2,
                select : function (event, ui) {
                    appManager.resetState();
                }
            });
        }                    
    }),
    refreshAppsView : Ember.View.extend({
        didInsertElement : function () {
            this.$("#refreshAppList").click(function (e){
                e.preventDefault();
                appManager.setView('appList', false);
                appManager.showAppList().done(function () {
                    appManager.checkForUpdates(); 
                });
            }); 
        }
    }),
    showAppsView : Ember.View.extend({
        didInsertElement : function () {
            this.$("#appManagerTabs").tabs();
        }
    }),
    editMetaView : Ember.View.extend({
        didInsertElement : function () {
            var that = this;
            var findInMeta = function (list, field) {
                var obj;
                list.some(function (v, i) {
                    if (v.name == field) {
                        obj = v;
                        return true;
                    }
                });
                return obj;
            };
            this.$("#resetMeta").click(function(e) {
                e.preventDefault();
                appManager.resetState();
                var appName = appManager.view.get('controller.appName');
                var controller = appManager.view.get('controller');
                Ember.set(controller, 'metaList', false);
                appManager.getMeta(appName);
            });
            this.$("#refreshFileList").click(function (e) {
                e.preventDefault();
                var appName = appManager.view.get('controller.appName');
                var metaList = appManager.view.get('controller.metaList')
                var data = { name : appName, mode : "fetchFileList" };
                var files; 
                appManager.doAjax(data, function (res) {
                    if (res.success && res.fileList) {
                        var files = findInMeta(metaList, 'files');
                        Ember.set(files, 'value', res.fileList);
                    }
                });
            });
            this.$("#saveMeta").click(function (e) {
                e.preventDefault();
                appManager.resetState();
                $(this).prop('disabled', true);
                var btn = this;
                var appName = appManager.view.get('controller.appName');
                data = appManager.getFormData(that);
                var metaList = appManager.view.get('controller.metaList');
                var files = findInMeta(metaList, 'files');
                data.files = files.value;
                data.name = appName;
                var form = { meta : JSON.stringify(data), mode : "saveMeta" };
                appManager.doAjax(form, function (res) {
                    if (res.success) {
                        var controller = appManager.view.get('controller');
                        Ember.set(controller, 'metaList', false);
                        appManager.getMeta(appName).done(function () {
                                appManager.setSuccess({
                                msg : "Successfully updated metafile!"
                            }); 
                        });
                    } else if (res.metaError) {
                        var metaList = appManager.view.get('controller.metaList');
                        var field = findInMeta(metaList, res.metaError.name);
                        Ember.set(field, 'error', res.metaError.error);
                    } 
                    $(btn).prop('disabled', false);
                });
            });
        }
    })

}
$(function($) {
    $.clira.commandFile({
        name : "app-manager",
        templatesFile : "/apps/app-manager/app-manager.hbs",
        prereqs : [
            "/external/selectric/selectric.js"
        ],
        commands : [
            {
                command : "show apps",
                help : "List all apps, update or install new apps",
                templateName : "app-manager",
                execute: function (view, cmd, parse, poss) {
                    appManager.view = view;
                    appManager.setMode('showApps');
                    appManager.setSource("localDisk");
                    
                    appManager.showAppList().done(function () {
                       appManager.checkForUpdates();
                    });
                }
            },
            {
                command : "install app",
                help : "Install an app via commandline",
                templateName : "app-manager",
                arguments : [
                    {
                        name : "url-or-path",
                        help : "App url or local path to app location",
                        type : "string",
                        nokeyword : true,
                        mandatory : true
                    }
                ],
                execute: function (view, cmd, parse, poss) {
                    appManager.view = view;
                    appManager.setMode("appUpdateCmdl");
                    var arg = poss.data["url-or-path"];
                    if (arg) {
                        var opts = {};
                        var a = document.createElement('a');
                        a.href = arg;
                        if (a.hostname == "github.com") {
                            opts.src = "github";
                            opts.url = arg;
                        } else if (a.hostname == window.location.hostname) {
                            opts.src = "localDisk";
                            opts.path = arg;
                        } else {
                            opts.src = "webServer";
                            opts.url = arg;
                        }
                        appManager.installApp(opts, function (res) {
                            if (res.error) {
                                appManager.setError(res.error);
                            } else if (res.success) {
                                $.clira.reloadApp(res.appName).done(function(){
                                    appManager.setSuccess({
                                        name : res.appName,
                                        result : 'installed'
                                    });
                                });
                            }
                        });
                        
                    }
                }
            },
            {
                command : "update app",
                help : "Update an app via commandline",
                templateName : "app-manager",
                arguments : [
                    {
                        name : "app-name",
                        help : "Name of application to be updated",
                        type : "string",
                        nokeyword : true,
                        mandatory : true
                    }
                ],
                execute: function (view, cmd, parse, poss) {
                    appManager.view = view;
                    appManager.setMode("appUpdateCmdl");
                    var appName = poss.data["app-name"];
                    appManager.setLoading("Updating '" + appName + "' app...");
                    appManager.updateApp(appName, function (res) {
                        if (res.success) {
                            if (res['up-to-date']) {
                                appManager.setSuccess({
                                    name : appName,
                                    uptodate : true,
                                }); 
                            } else {
                                $.clira.reloadApp(appName).done(function(){
                                    appManager.setSuccess({
                                        name : appName,
                                        result : 'updated',
                                        version : res['new-version']
                                    });
                                });
                            }
                        } else if (res.error) {
                            appManager.setError(res.error);
                        }
                    });
                }
            },
            {
                command : "edit metafile",
                help : "Create or edit an application metafile",
                templateName : "app-manager",
                arguments : [
                    {
                        name : "app-name",
                        help : "Update metafile belonging to app-name",
                        type : "string",
                        nokeyword : true,
                        mandatory : true
                    }
                ],
                execute: function (view, cmd, parse, poss) {
                    appManager.view = view;
                    appManager.setMode("editMeta");
                    var appName = poss.data["app-name"];
                    appManager.getMeta(appName);
                }
            }
        ] 
    });
});

File: web/apps/app-manager/py/app-manager.py
#CLIRA app installation core routines.

import urllib2, cgi, os, json, shutil, urlparse, base64, cgitb, sys, traceback
import time
from itertools import izip_longest

"""
Main wrapper to call various app management  routines depending on params 
passed
"""
def manageApps () :

    #extract the form data
    form = cgi.FieldStorage()
    src = form.getvalue('src')
    mode = form.getvalue('mode')

    if mode == "listApps" :
        res = getAppList()
        if res :
            updateResponse('appList', res)
            send("success", True)
    elif mode == "checkAppUpdate" :
        res = getUpdateInfo(form.getvalue('url'), form.getvalue('version'))
        if res : 
            updateResponse('updateInfo', res)
            send("success", True)
    elif mode == "fetchFileList" :
        filelist = getAppFileList(form.getvalue('name'))
        if filelist :
            updateResponse('fileList', filelist)
            send("success", True)
    elif mode == "getMeta" and getMeta(form.getvalue('name')) :
        send("success", True)
    elif mode == "saveMeta" and saveMeta(form.getvalue('meta')) :
        send("success", True)
    elif mode == "updateApp" and updateApp(form.getvalue('name')):
        send("success", True)
    elif src == "localDisk" and installFromDisk(form) :
        send("success", True)
    elif src == "github" and installFromGithub(form):
        send("success", True)
    elif src == "webServer" and installFromWebServer(form):
        send("success", True)


"""
List installed apps and check for update
"""
def getAppList () :
    appDir = appDirPath()
    appList = []
    dirs = os.listdir(appDir)
    for d in dirs :
        app = {}
        if not d.startswith(('.', '..')) and os.path.isdir(appDir + d):
            app['name'] = d
            metapath = appDir + d + "/" + d + ".meta"
            if not os.path.isfile(metapath) :
                app['meta'] = False
                appList.append(app)
                continue
            try : 
                with open(metapath) as fh :
                    meta = json.load(fh)
                    ret = validateMetaData(meta)
                    if 'error' in ret :
                        app['meta-error'] = ret['error']
                        app['meta'] = meta
                        appList.append(app)
                        continue
                    if 'app-meta-url' not in meta :
                        app['meta-error'] = "No app Update URL"
                    fh.close()
            except (ValueError, IOError) as e :
                app['meta'] = False
                app['meta-error'] = "Failed to load meta file for '" + d + "' : " + e(str)
                appList.append(app)
                continue
            app['meta'] = meta
            appList.append(app)
    return appList

def updateApp (appName) :
    path = appDirPath(appName) + appName + '.meta'
    if not os.path.exists(path) :
        error("Cannot update '" + appName + "'. App not installed or meta file is missing.")
        return False
    meta = getAppMetaData('fileDisk', path)
    if not meta :
        return False
    if 'app-meta-url' not in meta :
        error("Cannot update '" + appName + "', missing 'app-meta-url' from meta file")
        return False
    url = meta['app-meta-url']
    info = getUpdateInfo(url, meta['version'])
    if not info:
        return False
    if not info['new-version'] :
        updateResponse('up-to-date', True)
        return True
    updateResponse('new-version', info['new-version'])
    if not githubInstallCommon(info['meta'], 'update') :
        shutil.rmtree(appDirPath("_" + appName), ignore_errors=True)
        return False
    
    return True

"""
Check if an app has a newer version available using the URL given by the client
"""
def getUpdateInfo (url, version) :
    updateInfo = {}
    if not url :
        updateInfo['error'] = "Missing url field"
        return updateInfo
    if urlparse.urlparse(url).netloc == "github.com" :
        src = "github"
    else : 
        src = None
    meta = getAppMetaData(src, url)
    if meta :
        if 'not-modified' in meta :
            updateInfo['new-version'] = False
        else :
            if compareVersions(version, meta['version']) == -1 :
                updateInfo['new-version'] = meta['version']
            else :
                updateInfo['new-version'] = False
            updateInfo['meta'] = meta
    else :
        return False
    return updateInfo

"""
Compare versions v1, v2 and return :
    -1 if v1 < v2
     1 if v1 > v2
     0 if v1 == v2
"""
def compareVersions (v1, v2) :
    ver1 = (int(n) for n in v1.split('.'))
    ver2 = (int(n) for n in v2.split('.'))
    for i, j in izip_longest(ver1, ver2, fillvalue=0) :
        if i > j :
            return 1
        elif i < j :
            return -1
    return 0

"""
Install an app from disk using the params
"""
def installFromDisk (form) : 
    path = form.getvalue('path')
    try : 
        meta = getAppMetaData('fileDisk', path)
        if not meta:
            return False
        updateResponse('appName', meta["name"])
        srcDir = os.path.dirname(path) + "/"
        destDir = appDirPath(meta["name"])
        files = meta["files"]

        #Create the necessary app directories
        rmAppDir(meta['name'])
        if not createAppDirs(destDir, files) : 
            return False

        #Copy all the app files to the dest
        if not copyAppFiles(srcDir, destDir, files):
            return False
    except (IOError, OSError, ValueError) as e : 
        error(str(e))
        return False
    return True

"""
Use the app meta data and fetch application files using the Github API. 
Download files directly into the destination directory. For each 'file' in the
app meta 'files' field, we create the corresponding Github API URL and fetch
the data. The file content is contained base64 encoded in the 'content' field.
Take that and write to dest location.
"""
def installFromGithub (form) :
    url = form.getvalue('url')
    if not url :
        url = form.getvalue('app-meta-url')
    if not url :
        error("No app src URL defined.")
        return False
    appName = getAppNameFromURL(url)
    if not appName :
        return False
    updateResponse('appName', appName)
    if appInstalled(appName) :
        updateResponse('app-exists' , True)
        error("'" + appName + "' is already installed.")
        return False
    meta = getAppMetaData("github", url)
    if not meta :
        return False
    if not githubInstallCommon(meta, 'install') :
        shutil.rmtree(appDirPath("_" + appName), ignore_errors=True)
        return False
    
    return True

"""
Common set of tasks while installing or updating an app from Github
"""
def githubInstallCommon (meta, mode) :
    files = meta["files"]
    appName = meta['name']
    destDir = appDirPath("_" + appName)

    if not createAppDirs(destDir, files):
        error("Failed to create app directories")
        return False

    #First write the meta file, this prevents an unnecessary api call later
    if not updateAppMetaFile(meta, True):
        return False

    dirUrl = getDirUrl(meta['app-meta-url'])
    if dirUrl[-1] != '/' :
        dirUrl += '/'

    for f in files :
        # Skip writing the meta file, its handled already
        if f.lower().endswith('.meta'):
            continue
        fUrl = getGithubApiUrl(dirUrl + f)
        if not fUrl :
            return False
        fdata = githubApiGetBlob(fUrl)
        if not fdata :
            return False
        if not writeFile(destDir + f, fdata) :
            return False
    if mode == "update" :
        rmAppDir(appName)
    os.rename(destDir, appDirPath(appName))
    return True 

"""
Install an app from a webserver URL
"""
def installFromWebServer (form) :
    url = form.getvalue('url')
    if not url :
        error("Invalid URL")
        return False
    appName = getAppNameFromURL(url)
    if not appName :
        return False
    updateResponse('appName', appName)
    meta = getAppMetaData("webServer", url)
    if not meta :
        return False
    destDir = appDirPath(meta['name'])
    files = meta['files']
    rmAppDir(appName)
    if not createAppDirs(destDir, files) :
        return False

    #First write the meta file, this prevents an unnecessary api call later
    if not updateAppMetaFile(meta, True):
        return False

    dirUrl = getDirUrl(url)
    if dirUrl[-1] != '/' :
        dirUrl += '/'
    for f in files :
        # Skip writing the meta file, its handled already
        if f.lower().endswith('.meta'):
            continue
        res = getURL(dirUrl + f)
        if ('code' or 'reason') in res :
            printHttpError(res, json_content=False)
            return False
        fdata = res['content']
        if not fdata or not writeFile(destDir + f, fdata) :
            return False
    return True
    
"""
Write 'data' to 'dest' file path 
"""
def writeFile (dest, data) :
    try :
        with open(dest, "w") as fh :
            fh.write(data)
            fh.close()
            return True
    except (IOError, OSError) as e : 
        error("Failed to write file to" + dest + " : " + e(str))
    return False

"""
Return the new app directory relative to where we are. This should be changed
to using a CLIRA env variable that stores the clira apps directory rather 
than using this ugly method.
"""
def appDirPath (dirname=None) : 
    if not dirname :
        return os.path.realpath("../../") + "/"
    return os.path.realpath("../../") + "/" + dirname + "/"

def appMetaFile(appName) :
    return appDirPath() + appName + "/" + appName + '.meta'

def getDirUrl (url) :
    o = urlparse.urlparse(url)
    return urlparse.urlunparse((o.scheme, o.netloc, os.path.dirname(o.path), 
        None, None, None))

def appInstalled (appName) :
    return os.path.exists(appDirPath(appName))

def getAppNameFromURL (url) :
    o = urlparse.urlparse(url)
    fname = os.path.basename(o.path)
    if fname.lower().endswith('.meta'):
        (appName, ext) = fname.split('.')
        if appName :
            return appName
    error("Cannot extract app name from URL. Make sure the URL points to an app.meta file")
    return False

def getAppVersion (appName) :
    appMeta = appDirPath() + appName + '/' +appName + '.meta'
    try :
        with open(appMeta) as fh :
            meta = json.load(fh)
            if 'version' in meta :
                return meta['version']
            fh.close()
    except (ValueError, IOError) as e :
        error(str(e))
    return False

def encodeUserData(user, password):
        return "Basic " + (user + ":" + password).encode("base64").rstrip()

"""
Convert a github.com URL to a api.github.com URL. Its easier to copy paste a
github.com URL from the browser. We simply parse its components and create an
equivalent api.github.com URL which allows us to fetch more info about the 
resource.
"""
def getGithubApiUrl (url) :
    try :
        o = urlparse.urlparse(url)
        if not o.netloc == "github.com":
            error("Enter a valid Github URL for the file")
            return False
        (owner, repo, ctype, branch, path) = o.path[1::].split('/', 4)
        if branch == 'master' :
            return "https://api.github.com/repos/{!s}/{!s}/contents/{!s}".format(owner, repo, path)
        else :
            return "https://api.github.com/repos/{!s}/{!s}/contents/{!s}?ref={!s}".format(owner, repo, path, branch)
    except ValueError as e :
        error("Failed to get Github API URL : " + str(e))
    return False

"""
Fetch data using a URL. dataType is json if json decoded object is required as 
result. username and password is optional and is used for HTTP auth.
"""
def getURL (url, dataType = None, username = None, password=None, header=None) :
    try :
        req = urllib2.Request(url)
        if dataType == "json" :
            req.add_header('Accept', 'application/json')
        else :
            req.add_header('Accept', 'text/plain')
        req.add_header("Content-type", "application/x-www-form-urlencoded")
        if username and password :
            req.add_header('Authorization', encodeUserData(username, password))
        if header :
            for k in header.keys():
                req.add_header(k, header[k])
        res = urllib2.urlopen(req)
        if dataType == "json" :
            return { 'content' : json.loads(res.read()),
                     'info' : dict(res.info())
                   }
        else :
            return { 'content' : res.read(),
                     'info' : dict(res.info())
                   }
    except urllib2.HTTPError as e :
        data = {}
        data['code'] = e.code
        data['reason'] = e.reason
        content = e.read()
        if content :
            data['content'] = content
        return data
    except urllib2.URLError as e :
        data = {}
        data['reason'] = str(e.reason)
        return data
    except Exception as e :
        error("Failed to fetch URL : " + str(e))

"""
Query the Github API using the url and return the file/blob content. Throw an
error if we don't see what we want.
"""
def githubApiGetBlob (url, user=None, password=None) :
    res = getURL(url, "json")
    if not res :
        return False
    if ('code' or 'reason') in res :
        printHttpError(res)
        return False
    res = res['content']
    if 'type' not in res or res['type'] != "file":
        error("URL does not point to a file : type => " + res["type"])
        return False
    if "content" not in res :
        error("Missing file content in response")
        return False
    try : 
        return base64.b64decode(res["content"])
    except ValueError as e:
        error("Failed to parse file content : " + str(e))
    return False


"""
Check if a resource has been modified using GITHUB's conditional access API.
If the resource has changed, return the resource else return False.
"""
def githubGetIfModified (url, lastModified) :
    header = {}
    if lastModified :
        header['If-Modified-Since'] = lastModified
    res = getURL(url, 'json', None, None, header)
    if ('code' or 'reason') in res :
        if res['code'] == 304 :
            return { 'not-modified' : True }
        else :
            printHttpError(res)
            return False
    return res

"""
Validate mandatory meta data fields
"""
def validateMetaData (meta) :
    fields = ('name', 'version', 'files')
    for f in fields :
        if f not in meta :
            return { 
                    "error" : "meta file missing mandatory field '" + f + "'",
                    "name" : f
                   }
    return {}

def updateAppMetaFile (data, temp=False):
    if temp :
        path = appDirPath() + "_" + data['name'] + "/" + data['name'] + ".meta"
    else :
        path = appMetaFile(data['name'])
    try :
        with open(path, 'w') as fh :
            json.dump(data, fh, indent=4)
            fh.close()
            return True
    except Exception as e :
        error("Failed to update app meta file : " + str(e))
    return False

def getLastModified (path) :
    try :
        if os.path.isfile(path) :
            mtime = time.gmtime(os.path.getmtime(path))
            return time.strftime("%a, %d %b %Y %H:%M:%S GMT", mtime)
    except Exception as e :
       error(str(e))
    return None


def getAppMetaData (src, url) :
    try :
        if src == "github" :
            url = getGithubApiUrl(url)
            if not url :
                return False
            appName = getAppNameFromURL(url)
            if not appName:
                return False
            path = appMetaFile(appName)
            lastModified = getLastModified(path)
            res = githubGetIfModified(url, lastModified)
            if not res :
                return False
            if 'not-modified' in res :
                return { 'not-modified' : True }
            info = res['info']
            res = res['content']
            if 'content' in res : 
                meta = json.loads(base64.b64decode(res["content"]))
            else :
                error("Failed to find meta file 'content' in response")
                return False

        elif src == "fileDisk" :
            if not os.path.isfile(url):
                error("Invalid file path : " + url)
                return False
            with open(url, 'r') as fh :
                meta = json.load(fh)
                fh.close()
        else :
            res = getURL(url, 'json')
            if ('reason' or 'code') in res :
                printHttpError(res, json_content=False)
                return False
            meta = res['content']

        if meta :
            ret = validateMetaData(meta)
            if 'error' in ret:
                error(ret['error'])
                return False
            else :
                return meta
    except ValueError as e : 
        error("Failed to parse meta file : " + str(e))
    except IOError as e :
        error("Failed to read meta file : " + str(e))
    return False 

def createAppDirs (destDir, files, ) :
    if not mkdirs(destDir):
        return False
    seen = {}
    for f in files :
        path = os.path.dirname(f)
        if not path or path in seen : continue
        seen[path] = True
        if not mkdirs(destDir + path) :
            return False
    return True

def rmAppDir(appName) :
    if appInstalled(appName) :
        shutil.rmtree(appDirPath(appName), ignore_errors=True)

def copyAppFiles (srcDir, destDir, files):
    try :
        for f in files :
            shutil.copy(srcDir + f, destDir + f)
    except (OSError, IOError) as e : 
        error("Error copying file " + f + " : " + str(e))
        return False
    return True

def printHttpError (data, json_content=True) :
    try :
        msg = None
        if 'reason' in data and not 'code' in data :
            error("HTTP error : " + data['reason'])
            return False
        if 'code' in data :
            if 'content' not in data :
                return False
            if json_content :
                content = json.loads(data['content'])
            else :
                content = data['content']
            if 'message' in content :
                msg = content['message']
            if 'documentation_url' in content :
                msg = msg + " More info : " + content['documentation_url']
            if msg :
                error("Github API error: " + msg)
            else :
                error("HTTP error : " + str(data['code']) + " " + data['reason'])
    except ValueError as e :
        error(str(e))

def mkdirs (path) : 
    try :
        if not os.path.exists(path) :
            os.makedirs(path)
    except OSError as e: 
        error(str(e))
        return False
    return True

def getAppFileList (appName) :
    fileList = []
    try : 
        appDir = appDirPath(appName)
        for root, dirs, files in os.walk(appDir):
            for f in files:
                path = os.path.join(root, f)
                path = path.replace(appDir, "")
                fileList.append(path) 
        return fileList
    except Exception as e:
        error(str(e))
    return False


def getMeta(appName) :
    if not appInstalled(appName) :
        error("No such app installed")
        return False
    if not os.path.isfile(appMetaFile(appName)) :
        meta = {}
        meta['name'] = appName
        meta['files'] = getAppFileList(appName)
        updateResponse('metaCreate', True)
        updateResponse('meta', meta)
        return True
    try :
        with open(appMetaFile(appName)) as fh :
            meta = json.load(fh)
            res = validateMetaData(meta)
            if 'error' in res :
                error(res['error'])
                return False
            fh.close()
            updateResponse('meta', meta)
            return True
    
    except Exception as e :
        error(str(e))
    return False

def saveMeta (metaStr) : 
    try : 
        meta = json.loads(metaStr)
        res = validateMetaData(meta)
        if 'error' in res :
            send('metaError', res)
            return False
        appMeta = meta['name'] + '.meta'
        for i, f in enumerate(meta['files']) :
            if f.find("(Will be created)") > -1 :
                meta['files'][i] = appMeta
                break
        if not updateAppMetaFile(meta) :
            return False
        return True
    except ValueError as e :
        error(e(str))
    return False

"""
Update the global response data
"""
def updateResponse (key, value) :
    global response
    response[key] = value

def printResponse () :
    global response
    print json.dumps(response)

def error (msg) :
    updateResponse('error', msg)
    printResponse()

def send (key, value) :
    updateResponse(key, value)
    printResponse()


#Main : Let's do this!
cgitb.enable()
print "Content-Type: application/json\n\n"

#Our global response data which is sent to the client
response = {}

try :
    manageApps()   
except : 
    #For unknown erros we simply send over the traceback to the client
    error(traceback.format_exc())
